backend/src/packages/shifts/shift.service.ts

```diff
@@ -0,0 +1,184 @@
+  }
+
+  public async findByDriverUserId(
+    driverId: number,
+  ): Promise<ShiftCreateResponseDto[]> {
+    const shifts = await this.shiftRepository.find({
+      driverId,
+    });
+
+    return shifts.length > 0
+      ? shifts.map((it) => ShiftEntity.initialize(it).toObject())
+      : [];
+  }
+
+  public async create({
+    body,
+    user,
+  }: {
+    body: ShiftCreateRequestDto;
+    user: UserEntityObjectWithGroupT;
+  }): Promise<ShiftCreateResponseDto> {
+    const canCreateShift = await this.checkAccessToShift({
+      driverId: body.driverId,
+      user,
+    });
+
+    if (!canCreateShift) {
+      throw new HttpError({
+        status: HttpCode.FORBIDDEN,
+        message: HttpMessage.NOT_ACCESS,
+      });
+    }
+
+    const shift = await this.shiftRepository.create(body);
+
+    return ShiftEntity.initialize(shift).toObject();
+  }
+
+  public async close({
+    body,
+    user,
+    params,
+  }: {
+    body: ShiftCloseRequestDto;
+    user: UserEntityObjectWithGroupT;
+    params: Pick<ShiftEntityT, 'id'>;
+  }): Promise<ShiftCreateResponseDto> {
+    const shift = await this.findByShiftId(params.id);
+
+    const canCloseShift = await this.checkAccessToShift({
+      driverId: shift.driverId,
+      user,
+    });
+
+    if (!canCloseShift) {
+      throw new HttpError({
+        status: HttpCode.FORBIDDEN,
+        message: HttpMessage.NOT_ACCESS,
+      });
+    }
+
+    if (shift.endDate !== null) {
+      throw new HttpError({
+        status: HttpCode.BAD_REQUEST,
+        message: HttpMessage.SHIFT_ALREADY_CLOSED,
+      });
+    }
+
+    const updShift = await this.shiftRepository.update({
+      id: params.id,
+      payload: body,
+    });
+
+    await this.delete(updShift.id);
+
+    return ShiftEntity.initialize(updShift).toObject();
+  }
+
+  public async update({
+    id,
+    payload,
+  }: {
+    id: number;
+    payload: Partial<ShiftEntityT>;
+  }): Promise<ShiftCreateResponseDto> {
+    const result = await this.shiftRepository.update({ id, payload });
+
+    return ShiftEntity.initialize(result).toObject();
+  }
+
+  public async delete(id: number): Promise<boolean> {
+    const shift = await this.shiftRepository.update({
+      id,
+      payload: { deletedAt: new Date() },
+    });
+
+    return Boolean(shift);
+  }
+
+  private async checkAccessToShift({
+    user,
+    driverId,
+  }: {
+    user: UserEntityObjectWithGroupT;
+    driverId: ShiftEntityT['driverId'];
+  }): Promise<boolean> {
+    if (user.group.key === UserGroupKey.BUSINESS) {
+      const business = await this.businessService.findByOwnerId(user.id);
+      const driver = await this.driverService.findByUserId(driverId);
+
+      if (!business || !driver || business.id !== driver.businessId) {
+        return false;
+      }
+    } else if (user.group.key === UserGroupKey.DRIVER) {
+      if (user.id !== driverId) {
+        return false;
+      }
+    } else {
+      return false;
+    }
+
+    return true;
+  }
```

Comment:
```suggestion
 
    const { group } = user;
    const { key: groupKey } = group;

    if (groupKey === UserGroupKey.BUSINESS) {
     const isDriverBelongedToBusiness = await checkisDriverBelongedToBusiness({businessId, driverId});

     return isDriverBelongedToBusiness;
    }

    if(groupKey === UserGroupKey.DRIVER) {
      return user.id === driverId;
    }

    return false;
```

Reply:
and naming - checkHasAccessToShift

Reply:
and can we make this helper checkisDriverBelongedToBusiness - with only one db call?

Reply:
I've refactored the method and the db call